---
title: "Autism spectrum disorder: understanding the impact of SNPs on biological pathways in the fetal and adult cortex"
author: "Evgeniia Golovina"
date: "02/02/2021"
output:
    html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# install.packages("pacman")
# load libraries
pacman::p_load(ggplot2, pander, dplyr, tidyr)

setwd("/Users/foffa/Desktop/genetic_regulation_in_ASD")
```

This is a reproducibility report for "Autism spectrum disorder: understanding the impact of SNPs on biological pathways in the fetal and adult cortex" study. 

Python (version 3.6.9), R (version 4.0.2) and RStudio (version 1.2.5033) were used for data processing, analysis and visualisation.  

1. Fetal brain cortical plate and germinal zone neuron Hi-C data are available via the [dbGaP](https://www.ncbi.nlm.nih.gov/gap/) (accession: phs001190.v1.p1). Adult dorsolateral prefrontal cortex cells Hi-C data is available on [GEO](https://www.ncbi.nlm.nih.gov/geo/) (accession: GSE87112).
2. Total RNA-seq and WGS datasets across GTEx v8 tissues are available via the [dbGaP](https://www.ncbi.nlm.nih.gov/gap/) (accession: phs000424.v8.p2). Total RNA-seq and genotyping datasets for fetal brain cortical tissue from 14-21 postconceptional weeks (PCWs) are available via the dbGaP (https://www.ncbi.nlm.nih.gov/gap/) (accession: phs001900.v1.p1).
3. Human genome build hg38 release 75 (GRCh38) (Homo_sapiens_assembly38_noALT_noHLA_noDecoy.fasta) was downloaded from [gs://gtex-resources](https://console.cloud.google.com/storage/browser/gtex-resources/references).
4. SNP genomic positions for genome GRCh38p7 build 151 were obtained from ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7.
5. Gene annotations for GENCODE v26 (gencode.v26.GRCh38.genes.gtf) were downloaded from [gs://gtex-resources](https://console.cloud.google.com/storage/browser/gtex-resources/references).
6. Roadmap Epigenomics Project 15-state ChromHMM models for adult dorsolateral prefrontal cortex (E073_15_coreMarks_hg38lift_mnemonics.bed.gz) and fetal brain (E081_15_coreMarks_hg38lift_mnemonics.bed.gz) were downloaded from https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/.
7. The STRING PPI network (version 11.0, protein.links.full.v11.0.txt.gz) was downloaded from https://string-db.org/.

### 1. Identification of significant spatial eQTL SNP-gene interactions using CoDeS3D

First, we run CoDeS3D pipeline to get regulatory interactions across fetal cortical tissue: `python codes3d/codes3d.py -s data/asd_snps/344_asd_snps.txt -o results/codes3d/fetal_cortex -n Cortical_plate_neurons_Won2016 Germinal_zone_neurons_Won2016 -t Fetal_Brain_Cortex_Walker2019`.
Then across adult cortical tissue: `python codes3d/codes3d.py -s data/asd_snps/344_asd_snps.txt -o results/codes3d/adult_cortex -n Dorsolateral_prefrontal_cortex_cells_Schmitt2016 -t Brain_Cortex`

```{r loading_codes3d_results}
# get number of ASD-associated GWAS snps present in both fetal and adult eqtl databases
asd_snps <- readLines("data/asd_snps/344_asd_snps.txt") # 344

# read significant interactions in fetal and adult cortical tissues
fetal <- read.table("results/codes3d/fetal_cortex/significant_eqtls.txt", header = TRUE, sep = "\t")
adult <- read.table("results/codes3d/adult_cortex/significant_eqtls.txt", header = TRUE, sep = "\t")

# get eQTLs in fetal and adult cortical tissues
fetal_eqtls <- unique(fetal$snp) # 80
adult_eqtls <- unique(adult$snp) # 58

# get eGenes in fetal and adult cortical tissues
fetal_egenes <- unique(fetal$gene) # 81
adult_egenes <- unique(adult$gene) # 44
```

### 2. Percentage of eQTL SNPs vs non-eQTL SNPs

```{r eqtl_vs_non-eQTL, fig.width=7, fig.height=7}
snp.df <- data.frame(
  eqtl_db = rep(c('Adult', 'Fetal'), each=1),
  snps = rep(c("eQTL","non-eQTL"), each=2),
  number = c(length(adult_eqtls), length(fetal_eqtls), length(asd_snps)-length(adult_eqtls),
             length(asd_snps)-length(fetal_eqtls)),
  percentage = c(round(length(adult_eqtls)/length(asd_snps)*100, 2),
                 round(length(fetal_eqtls)/length(asd_snps)*100, 2), 
                 round((length(asd_snps)-length(adult_eqtls))/length(asd_snps)*100, 2),
                 round((length(asd_snps)-length(fetal_eqtls))/length(asd_snps)*100, 2)))

#pdf("figures/eQTLs_vs_non-eQTLs.pdf", width = 8, height = 9)
ggplot(snp.df, aes(x = snps, y = percentage, fill = eqtl_db)) +
    geom_bar(stat="identity", position = "dodge") + 
    theme_classic() +
    theme(plot.title = element_blank(),
          axis.title.x = element_blank(),
          legend.text=element_text(size=20),
          legend.title=element_blank(),
          legend.position = "bottom",
          legend.direction = "horizontal",
          axis.text=element_text(size=20, colour = "black"),
          axis.title=element_text(size=24, colour = "black"),
          strip.text.x = element_text(size = 19, colour = "black")) +
    geom_text(aes(y = percentage, label = paste0("n=", number)),
              position=position_dodge(width=0.9), vjust=-0.25, color = "black", 
              size = 8, fontface = 'italic') +
    scale_fill_manual(values=c("#92278F" ,"#009444")) +
    labs(y = "Percentage") #+
    #facet_grid(.~eQTL_db)
#dev.off()
```

We used Fisher’s exact test to check if the proportion of eqtls is significantly different in fetal and adult groups.

```{r fishers_exact_test}
# The Fisher’s exact test is used when the total sample size is less than 1000.
eqtl.df <- data.frame(tissue=c('Adult', 'Fetal'), 
                      eqtl=c(length(adult_eqtls), length(fetal_eqtls)),
                      non_eqtl=c(length(asd_snps)-length(adult_eqtls),
                                 length(asd_snps)-length(fetal_eqtls)))
rownames(eqtl.df) <- eqtl.df[,1]; eqtl.df[,1] <- NULL
f_test <- fisher.test(eqtl.df) # p-value = 0.04531
```

The proportion of eQTLs is significantly different in two groups with a p-value = `r pander(f_test$p.value)` (less than the significance level alpha = 0.05).  

Thirty ASD-associated eQTLs were shared between fetal and adult cortical tissues.

```{r shared_eqtls_and_non-eqtls, fig.width=7, fig.height=7}
shared_eqtls <- intersect(fetal_eqtls, adult_eqtls) # 30
shared_noneqtls <- intersect(asd_snps[!(asd_snps %in% fetal_eqtls)],
                             asd_snps[!(asd_snps %in% adult_eqtls)]) # 236

shared.df <- data.frame(
  snps = rep(c("eQTL","non-eQTL")),
  number = c(length(shared_eqtls), length(shared_noneqtls)),
  percentage = c(round(length(shared_eqtls)/length(asd_snps)*100, 2), 
                 round(length(shared_noneqtls)/length(asd_snps)*100, 2)))

#pdf("figures/shared_eQTL_and_non-eQTL.pdf", width = 5, height = 9)
ggplot(shared.df, aes(x = snps, y = percentage, fill = "#A9A9A9")) +
    geom_bar(stat="identity", position = "dodge") + 
    theme_classic() +
    theme(plot.title = element_blank(),
          axis.title.x = element_blank(),
          legend.text=element_text(size=20),
          legend.title=element_blank(),
          legend.position = "none",
          axis.text=element_text(size=20, colour = "black"),
          axis.title=element_text(size=24, colour = "black"),
          strip.text.x = element_text(size = 19, colour = "black")) +
    geom_text(aes(y = percentage, label = paste0("n=", number)),
              position=position_dodge(width=0.9), vjust=-0.25, color = "black", 
              size = 8, fontface = 'italic') +
    labs(y = "Percentage") + 
    scale_fill_manual(values=c("#A9A9A9"))
#dev.off()
```

### 3. Functional annotation of eQTLs associated with ASD

We used [wANNOVAR tool](http://wannovar.wglab.org/) to obtain information about the locus eQTLs tagged. First, we extracted the funnctional annotation for ASD-associated eQTLs: `cut -f1,2,3,6,132 results/wannovar/asd_genome_summary.txt | sort -u > results/wannovar/asd_fun_snp_ann.txt`

```{r wANNOVAR, fig.width=7, fig.height=5}
# Calculate percentage of functionally annotated SNPs
cal_ann <- function(phe, total) {
    df <- data.frame(matrix(ncol=3, nrow=0))
    colnames(df)<- c("annotation","number", "percent")
    for (i in 1:length(phe)){
        down <- nrow(phe[which(phe$Annotation=="downstream"), ])
        df[nrow(df) + 1,] = list(annotation = "downstream", number = down, percent = down/total*100)
        ex <- nrow(phe[which(phe$Annotation=="exonic"), ])
        df[nrow(df) + 1,] = list(annotation = "exonic", number = ex, percent = ex/total*100)
        inter <- nrow(phe[which(phe$Annotation=="intergenic"), ])
        df[nrow(df) + 1,] = list(annotation = "intergenic", number = inter, percent = inter/total*100)
        intro <- nrow(phe[which(phe$Annotation=="intronic"), ])
        df[nrow(df) + 1,] = list(annotation = "intronic", number = intro, percent = intro/total*100)
        ncRNA_ex <- nrow(phe[which(phe$Annotation=="ncRNA_exonic"), ])
        df[nrow(df) + 1,] = list(annotation = "ncRNA_exonic", number = ncRNA_ex, percent = ncRNA_ex/total*100)
        ncRNA_in <- nrow(phe[which(phe$Annotation=="ncRNA_intronic"), ])
        df[nrow(df) + 1,] = list(annotation = "ncRNA_intronic", number = ncRNA_in, percent = ncRNA_in/total*100)
        up <- nrow(phe[which(phe$Annotation=="upstream"), ])
        df[nrow(df) + 1,] = list(annotation = "upstream", number = up, percent = up/total*100)
        UTR3 <- nrow(phe[which(phe$Annotation=="UTR3"), ])
        df[nrow(df) + 1,] = list(annotation = "UTR3", number = UTR3, percent = UTR3/total*100)
        }
    df[!duplicated(df), ]
}

asd_ann <- read.table("results/wannovar/asd_fun_snp_ann.txt", sep = "\t", header=TRUE) # wANNOVAR annotation for all ASD-associated snps
asd_ann_fetal <- subset(asd_ann, asd_ann$SNP %in% fetal_eqtls) # 80
asd_ann_adult <- subset(asd_ann, asd_ann$SNP %in% adult_eqtls) # 58

asd_fun_ann_fetal <- cal_ann(asd_ann_fetal, length(fetal_eqtls))
asd_fun_ann_adult <- cal_ann(asd_ann_adult, length(adult_eqtls))

# Fetal eQTLs
f_number <- c(asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="downstream",][,2],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="exonic",][,2],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="intergenic",][,2],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="intronic",][,2],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="ncRNA_intronic",][,2],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="ncRNA_exonic",][,2],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="UTR3",][,2])
f_percent <- c(asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="downstream",][,3],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="exonic",][,3],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="intergenic",][,3],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="intronic",][,3],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="ncRNA_intronic",][,3],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="ncRNA_exonic",][,3],
              asd_fun_ann_fetal[asd_fun_ann_fetal$annotation=="UTR3",][,3])

fetal_eqtls.df <- data.frame(
    annotation = c("downstream", "exonic", "intergenic", "intronic", "ncRNA_intronic",
                 "ncRNA_exonic", "UTR3"),
    number = f_number,
    percent = f_percent)

#pdf("figures/asd_fetal_fun_ann.pdf", width = 6, height = 4)
ggplot(fetal_eqtls.df, aes(x = annotation, y = percent, fill = "#009444")) +
    geom_bar(stat="identity", position = "dodge") + 
    theme_classic() +
    theme(plot.title = element_blank(),
          axis.title.y = element_blank(),
          legend.position = "none",
          axis.text=element_text(size=14, colour = "black"),
          axis.title=element_text(size=16, colour = "black")) +
    scale_fill_manual(values = "#009444") +
    geom_text(aes(y = percent, label = paste0(round(percent, digits=2), "%")),
              position=position_dodge(width=0.9), vjust = 0.25, hjust = -0.25, color = "black",
              size = 3.5) + 
    scale_y_continuous(limits = c(0, 55)) +
    ylab("Percentage") + 
    coord_flip()
#dev.off()

# Adult eQTLs
a_number <- c(asd_fun_ann_adult[asd_fun_ann_adult$annotation=="downstream",][,2],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="exonic",][,2],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="intergenic",][,2],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="intronic",][,2],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="ncRNA_intronic",][,2],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="ncRNA_exonic",][,2],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="UTR3",][,2])
a_percent <- c(asd_fun_ann_adult[asd_fun_ann_adult$annotation=="downstream",][,3],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="exonic",][,3],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="intergenic",][,3],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="intronic",][,3],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="ncRNA_intronic",][,3],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="ncRNA_exonic",][,3],
              asd_fun_ann_adult[asd_fun_ann_adult$annotation=="UTR3",][,3])

adult_eqtls.df <- data.frame(
    annotation = c("downstream", "exonic", "intergenic", "intronic", "ncRNA_intronic",
                 "ncRNA_exonic", "UTR3"),
    number = a_number,
    percent = a_percent)

#pdf("figures/asd_adult_fun_ann.pdf", width = 6, height = 4)
ggplot(adult_eqtls.df, aes(x = annotation, y = percent, fill = "#92278F")) +
    geom_bar(stat="identity", position = "dodge") + 
    theme_classic() +
    theme(plot.title = element_blank(),
          axis.title.y = element_blank(),
          legend.position = "none",
          axis.text=element_text(size=14, colour = "black"),
          axis.title=element_text(size=16, colour = "black")) +
    scale_fill_manual(values = "#92278F") +
    geom_text(aes(y = percent, label = paste0(round(percent, digits=2), "%")),
              position=position_dodge(width=0.9), vjust = 0.25, hjust = -0.25, color = "black",
              size = 3.5) + 
    scale_y_continuous(limits = c(0, 55)) +
    ylab("Percentage") + 
    coord_flip()
#dev.off()
```

Enrichment of fetal and adult eQTLs within transcription factor binding sites was determined using [SNP2TFBS](https://ccg.epfl.ch//snp2tfbs/).

```{r SNP2TFBS, fig.width=9, fig.height=5}
snp2tfbs <- read.table("results/snp2tfbs/match_output_20913.txt", header = FALSE)
fetal_snp2tfbs <- snp2tfbs[snp2tfbs$V7 %in% fetal_eqtls, ]
adult_snp2tfbs <- snp2tfbs[snp2tfbs$V7 %in% adult_eqtls, ]
shared_snp2tfbs <- snp2tfbs[snp2tfbs$V7 %in% shared_eqtls, ]

# split the V6 column by ";"
fetal_snp2tfbs <- separate(fetal_snp2tfbs, col=V6, into = c("MATCH", "TF", "ScoreDiff"), sep = ";")
adult_snp2tfbs <- separate(adult_snp2tfbs, col=V6, into = c("MATCH", "TF", "ScoreDiff"), sep = ";")
shared_snp2tfbs <- separate(shared_snp2tfbs, col=V6, into = c("MATCH", "TF", "ScoreDiff"), sep = ";")

# create a dataframe with number of affected TFBSs by fetal, adult or shared eqtls
snp2tfbs.df <- data.frame(
    eqtl_db = rep(c('Adult', 'Fetal', 'Shared'), each=5),
    tfbs = rep(1:5, 3),
    snps = c(nrow(adult_snp2tfbs[adult_snp2tfbs$MATCH=="MATCH=1",]),
             nrow(adult_snp2tfbs[adult_snp2tfbs$MATCH=="MATCH=2",]),
             nrow(adult_snp2tfbs[adult_snp2tfbs$MATCH=="MATCH=3",]),
             nrow(adult_snp2tfbs[adult_snp2tfbs$MATCH=="MATCH=4",]),
             nrow(adult_snp2tfbs[adult_snp2tfbs$MATCH=="MATCH=5",]),
             nrow(fetal_snp2tfbs[fetal_snp2tfbs$MATCH=="MATCH=1",]),
             nrow(fetal_snp2tfbs[fetal_snp2tfbs$MATCH=="MATCH=2",]),
             nrow(fetal_snp2tfbs[fetal_snp2tfbs$MATCH=="MATCH=3",]),
             nrow(fetal_snp2tfbs[fetal_snp2tfbs$MATCH=="MATCH=4",]),
             nrow(fetal_snp2tfbs[fetal_snp2tfbs$MATCH=="MATCH=5",]),
             nrow(shared_snp2tfbs[shared_snp2tfbs$MATCH=="MATCH=1",]),
             nrow(shared_snp2tfbs[shared_snp2tfbs$MATCH=="MATCH=2",]),
             nrow(shared_snp2tfbs[shared_snp2tfbs$MATCH=="MATCH=3",]),
             nrow(shared_snp2tfbs[shared_snp2tfbs$MATCH=="MATCH=4",]),
             nrow(shared_snp2tfbs[shared_snp2tfbs$MATCH=="MATCH=5",])))

#pdf("figures/fun_annotation/snp2tfbs_geombar.pdf", width = 10, height = 6)
ggplot(data = snp2tfbs.df, aes(x=tfbs, y=snps, fill=eqtl_db)) +
    geom_bar(stat="identity", position = "dodge") +
    theme_classic() +
    theme(plot.title = element_blank(),
          legend.position = "none",
          axis.text=element_text(size=20, colour = "black"),
          axis.title=element_text(size=24, colour = "black"),
          strip.text.x = element_text(size = 18, colour = "black")) +
    labs(x = "Number of TFBS affected", y = "Number of SNPs", color=" ") +
    scale_fill_manual(values=c("#92278F" ,"#009444", "darkgray")) +
    facet_grid(.~eqtl_db) #+
    #coord_flip()
#dev.off()
```


